File: cpp_version/geometry/matrix4.cpp
#include "pch.h"
#include "matrix4.h"
#include "vec3.h"

Matrix4::Matrix4() {
    for (int i = 0; i < 16; i++) {
        m[i] = 0.0f;
    }
    m[0] = 1.0f;
    m[5] = 1.0f;
    m[10] = 1.0f;
    m[15] = 1.0f;
}

Matrix4 Matrix4::identity() {
    return Matrix4();
}

Matrix4 Matrix4::translation(float x, float y, float z) {
    Matrix4 result;
    result.m[12] = x;
    result.m[13] = y;
    result.m[14] = z;
    return result;
}

Matrix4 Matrix4::translation(const Vec3& position) {
    return translation(
        (float)position.getX(),
        (float)position.getY(),
        (float)position.getZ()
    );
}

Matrix4 Matrix4::rotation(float x, float y, float z, float angleRadians) {
    float length = sqrtf(x * x + y * y + z * z);
    if (length == 0.0f) {
        return Matrix4::identity();
    }
    x /= length;
    y /= length;
    z /= length;

    float c = cosf(angleRadians);
    float s = sinf(angleRadians);
    float t = 1.0f - c;

    Matrix4 result;
    result.m[0] = t * x * x + c;
    result.m[1] = t * x * y + s * z;
    result.m[2] = t * x * z - s * y;

    result.m[4] = t * x * y - s * z;
    result.m[5] = t * y * y + c;
    result.m[6] = t * y * z + s * x;

    result.m[8] = t * x * z + s * y;
    result.m[9] = t * y * z - s * x;
    result.m[10] = t * z * z + c;
    return result;
}

Matrix4 Matrix4::scale(float x, float y, float z) {
    Matrix4 result;
    result.m[0] = x;
    result.m[5] = y;
    result.m[10] = z;
    return result;
}

Matrix4 Matrix4::operator*(const Matrix4& other) const {
    Matrix4 result;
    for (int col = 0; col < 4; col++) {
        for (int row = 0; row < 4; row++) {
            float sum = 0.0f;
            for (int k = 0; k < 4; k++) {
                sum += m[k * 4 + row] * other.m[col * 4 + k];
            }
            result.m[col * 4 + row] = sum;
        }
    }
    return result;
}

void Matrix4::copyTo(float out[16]) const {
    for (int i = 0; i < 16; i++) {
        out[i] = m[i];
    }
}

File: unity pmugg/Assets/GraphGrammar/Editor/Matrix4.cs
using System.Runtime.InteropServices;
using UnityEngine;

namespace Grammar {
    // Must match cpp_version/geometry/matrix4.h. Column-major: element (row, col) is m[col * 4 + row].
    [StructLayout(LayoutKind.Sequential)]
    public struct Matrix4 {
        public float m0, m1, m2, m3;
        public float m4, m5, m6, m7;
        public float m8, m9, m10, m11;
        public float m12, m13, m14, m15;

        public static readonly Matrix4 Identity = new Matrix4 {
            m0 = 1f,
            m5 = 1f,
            m10 = 1f,
            m15 = 1f,
        };

        // Swaps grammar Y/Z into Unity's coordinate system.
        public static readonly Matrix4 GrammarToUnity = new Matrix4 {
            m0 = 1f,
            m5 = 0f,
            m6 = 1f,
            m7 = 0f,
            m9 = 1f,
            m10 = 0f,
            m15 = 1f,
        };

        public static Matrix4 operator *(Matrix4 left, Matrix4 right) {
            Matrix4 result = default;
            for (int col = 0; col < 4; col++) {
                for (int row = 0; row < 4; row++) {
                    float sum = 0f;
                    for (int k = 0; k < 4; k++) {
                        sum += left.At(k, row) * right.At(col, k);
                    }
                    result.Set(col, row, sum);
                }
            }
            return result;
        }

        public void Decompose(out Vector3 position, out Quaternion rotation, out Vector3 scale) {
            position = new Vector3(m12, m13, m14);

            Vector3 column0 = new Vector3(m0, m1, m2);
            Vector3 column1 = new Vector3(m4, m5, m6);
            Vector3 column2 = new Vector3(m8, m9, m10);

            scale = new Vector3(column0.magnitude, column1.magnitude, column2.magnitude);

            if (scale.x > 1e-6f) {
                column0 /= scale.x;
            }
            if (scale.y > 1e-6f) {
                column1 /= scale.y;
            }
            if (scale.z > 1e-6f) {
                column2 /= scale.z;
            }

            var rotationMatrix = new Matrix4x4(column0, column1, column2, new Vector4(0f, 0f, 0f, 1f));
            rotation = rotationMatrix.rotation;
        }

        public void ApplyTo(Transform target) {
            Matrix4 unityMatrix = GrammarToUnity * this * GrammarToUnity;
            unityMatrix.Decompose(out Vector3 position, out Quaternion rotation, out Vector3 scale);
            target.localPosition = position;
            target.localRotation = rotation;
            target.localScale = scale;
        }

        float At(int col, int row) {
            return this[col * 4 + row];
        }

        void Set(int col, int row, float value) {
            this[col * 4 + row] = value;
        }

        float this[int index] {
            get {
                switch (index) {
                    case 0: return m0;
                    case 1: return m1;
                    case 2: return m2;
                    case 3: return m3;
                    case 4: return m4;
                    case 5: return m5;
                    case 6: return m6;
                    case 7: return m7;
                    case 8: return m8;
                    case 9: return m9;
                    case 10: return m10;
                    case 11: return m11;
                    case 12: return m12;
                    case 13: return m13;
                    case 14: return m14;
                    case 15: return m15;
                    default: return 0f;
                }
            }
            set {
                switch (index) {
                    case 0: m0 = value; break;
                    case 1: m1 = value; break;
                    case 2: m2 = value; break;
                    case 3: m3 = value; break;
                    case 4: m4 = value; break;
                    case 5: m5 = value; break;
                    case 6: m6 = value; break;
                    case 7: m7 = value; break;
                    case 8: m8 = value; break;
                    case 9: m9 = value; break;
                    case 10: m10 = value; break;
                    case 11: m11 = value; break;
                    case 12: m12 = value; break;
                    case 13: m13 = value; break;
                    case 14: m14 = value; break;
                    case 15: m15 = value; break;
                }
            }
        }
    }
}
